tests/datatypes/test_str.py

```diff
@@ -195,6 +197,378 @@ def test_startswith_tuple(self):
                 print('BeeWare starts with either e or B')
          """)
 
+class FormatTests(TranspileTestCase):
+
+        alternate = ('#',
+                     ''
+         )
+
+        length_modifiers = (
+            'h',
+            'L',
+            'l',
+            ''
+        )
+
+        conversion_flags = (
+            'd',
+            'i',
+            'o',
+            'u',
+            'x',
+            'X',
+            'e',
+            'E',
+            'f',
+            'F',
+            'g',
+            'G',
+            'r',
+            's'
+        )
+
+        args = (
+            '"s"',
+            '"spam"',
+            '"5"',
+            5,
+            -5,
+            5.0,
+            -5.0,
+            0.5,
+            0.50,
+            0.000005,
+            0.000000000000000000005,
+            -0.5,
+            -0.000005,
+            -0.000000000000000000005,
+            500000,
+            -500000,
+            500000000000000000000,
+            -500000000000000000000,
+            1361129467683753853853498429727072845824,
+            1361129467683753853853498429727072845824
+        )
+
+        template = """
+                print('>>> "format this: %{spec}" % {arg}')
+                try:
+                    print('format this: %{spec}' % {arg})
+                except (ValueError, TypeError, OverflowError) as err:
+                    print(err)
+                print('Done.')
+                """
+
+        @transforms(
+            js_bool=False,
+            decimal=False,
+            float_exp=False,
+            memory_ref=False
+        )
+        def test_basic(self, js_cleaner, py_cleaner):
+            combinations = (product(self.alternate, self.conversion_flags, self.args))
+            tests = ''.join([adjust(
+                self.template
+                    .format(
+                    spec=comb[0] + comb[1], arg=comb[2])
+            ) for comb in combinations]
+                            )
+
+            self.assertCodeExecution(tests, js_cleaner=js_cleaner, py_cleaner=py_cleaner)
+
+        @transforms(
+            js_bool=False,
+            decimal=False,
+            float_exp=False
+        )
+        def test_c_conversion(self, js_cleaner, py_cleaner):
+            """tests for c character conversion
+            in C Python there is an upper bound to what int or float can be provided
+            and this is platform specific. currently, Batavia is not enforcing any
+            kind of upper bound.
+            """
+            values = [
+                32,
+                100,
+                -1,
+                '"s"',
+                '"spam"'
+            ]
```

Comment:
For internal purposes, it would be helpful to annotate why these particular cases are interesting (i.e, space, 'd', and three invalid values)